rubix-core/src/main/java/com/qubole/rubix/core/CachingInputStream.java

```diff
@@ -230,11 +234,39 @@ int readFullyDirect(byte[] buffer, int offset, int length)
     return nread;
   }
 
+  @Override
+  public int read(long position, byte[] buffer, int offset, int length)
+          throws IOException
+  {
+    long oldPos = getPos();
+    try {
+      seek(position);
+      return readInternal(buffer, offset, length);
+    }
+    catch (InterruptedException e) {
+      throw Throwables.propagate(e);
+    }
+    catch (Exception e) {
+      log.error(String.format("Failed to read from rubix for file %s position %d length %d. Falling back to remote", remotePath, nextReadPosition, length), e);
+      try {
+        getParentDataInputStream().readFully(position, buffer, offset, length);
+      }
+      catch (Exception ex) {
+        log.error(format("Error reading %d bytes directly from remote for file %s at position %d", length, remotePath, nextReadPosition), ex);
```

Comment:
wouldnt this exception be part of engine error/logs? We would do double logging with this

Reply:
ya, i've removed it now